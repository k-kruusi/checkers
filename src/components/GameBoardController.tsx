import React, { useCallback, useState } from 'react';
import { useCheckers, useNextMove } from '../hooks';
import { Coord, MoveResult, Piece } from '../schema';
import { theme } from '../theme';
import { ActionType } from '../reducer';
import { TileContainer } from './TileContainer';
import { BannerController } from './BannerController';
import { ResetButton } from './ResetButton';



const gameBoardStyle: React.CSSProperties = {
  display: 'grid',
  gridTemplateColumns: 'repeat(8, auto)',
  gap: 0,
  border: `4px solid ${theme.colors.gold}`,
  boxShadow: '0 4px 8px rgba(0, 0, 0, 0.2)',
  borderRadius: 4,
  padding: 4,
}

const gameBoardContainerStyle: React.CSSProperties = {
  position: 'relative',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  padding: 20
}

export const GameBoardController: React.FC = () => {
  const { state, dispatch } = useCheckers();
  const { board, turn } = state;
  const { inspect, potentials } = useNextMove();
  const [isDragging, setIsDragging] = useState<Coord | null>(null);

  const handleDrop = useCallback((from: Coord, to: Coord) => {
    const chosen = potentials.find((p) => p.coord.x === to.x && p.coord.y === to.y);
    if (chosen) {
      dispatch({
        type: ActionType.MOVE_PIECE,
        from,
        to,
        piece: board[from.y][from.x],
        potentials: [...potentials],
        result: chosen.didJump ? MoveResult.Jump : MoveResult.Shift,
      });
    }
    else {
      console.log('not in chosen');
    }
    setIsDragging(null);
  }, [potentials, dispatch]);

  const handleDragEnd = useCallback(() => {
    setIsDragging(null);
  }, [setIsDragging]);

  const handleHover = useCallback((piece: Piece, coord: Coord) => {
    !isDragging && inspect({ piece, coord });
  }, [isDragging, inspect]);

  const handleDragStart = useCallback((piece: Piece, coord: Coord) => {
    setIsDragging(coord);
    inspect({ piece, coord });
  }, [setIsDragging, inspect]);

  const getTileColor = useCallback((x: number, y: number) => {
    const isOddRow = y % 2 === 0 ? false : true;
    const isOddTile = x % 2 === 0 ? false : true;
    const defaultTileBackground = isOddRow ?
      isOddTile ?
        theme.colors.ivory :
        theme.colors.grey :
      !isOddTile ?
        theme.colors.ivory :
        theme.colors.grey;

    return potentials.some(p => p.coord.x === x && p.coord.y === y) ?
      theme.colors.blue :
      defaultTileBackground;
  }, [potentials]);

  return (
    <>
      <div style={gameBoardContainerStyle} onDragEnd={handleDragEnd}>
        <div style={gameBoardStyle}>
          {
            board.map((row, rowIndex) => (
              row.map((cell, colIndex) => {
                return (
                  <div key={rowIndex + "-" + colIndex} >
                    <TileContainer
                      piece={cell}
                      x={colIndex}
                      y={rowIndex}
                      turn={turn}
                      handleDragStart={handleDragStart}
                      handleDrop={handleDrop}
                      handleDragEnd={handleDragEnd}
                      handleHover={handleHover}
                      getTileColor={getTileColor}
                      isDragging={isDragging}
                    />
                  </div>
                )
              })
            ))
          }
        </div>
        <BannerController />
      </div>
      <ResetButton />
    </>
  );
};
