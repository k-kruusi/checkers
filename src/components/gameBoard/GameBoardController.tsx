import React, { useCallback, useState } from 'react';
import { useCheckers, useNextMove } from '../../hooks';
import { Coord, MoveResult, Piece } from '../../schema';
import { theme } from '../../theme';
import { ActionType } from '../../reducer';
import { TileColor, TileController } from './TileController';

const gameBoardStyle: React.CSSProperties = {
  display: 'grid',
  gridTemplateColumns: 'repeat(8, auto)',
  gap: 0,
  border: `4px solid ${theme.colors.gold}`,
  boxShadow: '0 4px 8px rgba(0, 0, 0, 0.2)',
  borderRadius: 4,
  padding: 4,
};

const gameBoardContainerStyle: React.CSSProperties = {
  position: 'relative',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  margin: "20px 20px 20px 20px",
};

// Big picture manages the state of the board, passes appropriate methods to the tiles
export const GameBoardController: React.FC = () => {
  const { state, dispatch } = useCheckers();
  const { board } = state;
  const { inspect, potentials, clear } = useNextMove();
  const [isDragging, setIsDragging] = useState<Coord | null>(null);

  const handleDrop = useCallback((from: Coord, to: Coord) => {
    const validMove = potentials.find((p) => p.coord.x === to.x && p.coord.y === to.y);
    if (!validMove) {
      console.log('Invalid move attempted.');
      setIsDragging(null);
      clear();
      return;
    }

    dispatch({
      type: ActionType.MOVE_PIECE,
      from,
      to,
      piece: board[from.y][from.x],
      potentials: [...potentials],
      result: validMove.didJump ? MoveResult.Jump : MoveResult.Shift,
      timestamp: new Date().toISOString(),
    });

    setIsDragging(null);
    clear();
  }, [potentials, board, dispatch, clear, setIsDragging]);

  const handleDragEnd = useCallback(() => {
    setIsDragging(null);
  }, [setIsDragging]);

  const handleHover = useCallback((piece: Piece, coord: Coord) => {
    !isDragging && inspect({ piece, coord });
  }, [isDragging, inspect]);

  const handleDragStart = useCallback((piece: Piece, coord: Coord) => {
    setIsDragging(coord);
    inspect({ piece, coord });
  }, [setIsDragging, inspect]);

  const getTileColor = useCallback((x: number, y: number): TileColor => {
    const isOddRow = y % 2 === 0 ? false : true;
    const isOddTile = x % 2 === 0 ? false : true;
    const defaultTileBackground = isOddRow ?
      isOddTile ?
        theme.colors.ivory :
        theme.colors.grey :
      !isOddTile ?
        theme.colors.ivory :
        theme.colors.grey;

    return potentials.some(p => p.coord.x === x && p.coord.y === y) ?
      { color: defaultTileBackground, highlight: theme.colors.blue } :
      { color: defaultTileBackground, highlight: null };
  }, [potentials]);

  return (
    <>
      <div style={gameBoardContainerStyle} onDragEnd={handleDragEnd}>
        <div style={gameBoardStyle}>
          {
            board.map((row, rowIndex) => (
              row.map((cell, colIndex) => {
                return (
                  <div key={rowIndex + '-' + colIndex} >
                    <TileController
                      piece={cell}
                      x={colIndex}
                      y={rowIndex}
                      handleDragStart={handleDragStart}
                      handleDrop={handleDrop}
                      handleDragEnd={handleDragEnd}
                      handleHover={handleHover}
                      getTileColor={getTileColor}
                      isDragging={isDragging}
                    />
                  </div>
                )
              })
            ))
          }
        </div>
      </div>
    </>
  );
};
