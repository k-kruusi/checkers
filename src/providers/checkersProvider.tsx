import { ReactNode, createContext, useMemo, useReducer } from 'react';
import { BoardState } from '../schema';
import { Action, reducer } from '../reducer';
import { Timer, cloneState } from '../utils';

export const CheckersContext = createContext<{
  state: BoardState;
  dispatch: React.Dispatch<Action>;
  blackTimer: Timer;
  redTimer: Timer;
} | null>(null);

export const CheckersProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, cloneState());

  const blackTimer = useMemo(() => new Timer(), []);
  const redTimer = useMemo(() => new Timer(), []);

  return (
    <CheckersContext.Provider value={{ state, dispatch, blackTimer, redTimer }}>
      {children}
    </CheckersContext.Provider>
  );
};
