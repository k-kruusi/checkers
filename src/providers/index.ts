export * from './checkersProvider';
export * from './computerPlayerProvider';
export * from './nextMoveInspectorProvider';