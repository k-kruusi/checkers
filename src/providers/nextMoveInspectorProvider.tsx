import { createContext, useState } from 'react';
import { Outcome, Piece, TileData } from '../schema';
import { useCheckers } from '../hooks';
import { calculatePotentialMoves } from '../utils';

export const NextMoveInspectorContext = createContext<{
  inspect: (maybe: TileData) => void;
  inspectAndReturn: (maybe: TileData) => Outcome[];
  potentials: Outcome[];
  clear: () => void;
} | null>(null);

export const NextMoveInspectorProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [focus, setFocus] = useState<TileData | null>(null);
  const [potentials, setPotentials] = useState<Outcome[]>([]);
  const { state } = useCheckers();

  const inspect = (fresh: TileData) => {
    if (focus &&
      focus.piece === fresh.piece &&
      focus.coord.x === fresh.coord.x &&
      focus.coord.y === fresh.coord.y) {
      // exit early already calculated for that tile.
      return;
    }
    // if you are hovering over an empty
    // flush the potentials
    if (fresh.piece === Piece.Empty) {
      setPotentials([]);
      setFocus(fresh);
      return;
    }

    try {
      const moves = calculatePotentialMoves(state.board, fresh.coord);
      setPotentials(moves);
      setFocus(fresh);
    } catch (e) {
      console.error('error calculating potential moves', e);
    }
  };

  const inspectAndReturn = (fresh: TileData): Outcome[] => {
    try {
      return calculatePotentialMoves(state.board, fresh.coord);
    } catch (e) {
      console.error('error inspecting possible moves');
      return [];
    }
  }

  const clear = () => setPotentials([]);

  return (
    <NextMoveInspectorContext.Provider value={{ inspect, inspectAndReturn, potentials, clear }}>
      {children}
    </NextMoveInspectorContext.Provider>
  );
};


