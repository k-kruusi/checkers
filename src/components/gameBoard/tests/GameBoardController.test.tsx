import { render } from '@testing-library/react';
import { GameBoardController } from '../GameBoardController';
import { CheckersProvider, NextMoveInspectorProvider } from '../../../providers';

describe('GameBoardController', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <CheckersProvider>
        <NextMoveInspectorProvider>
          <GameBoardController />
        </NextMoveInspectorProvider>
      </CheckersProvider>);
    expect(asFragment()).toMatchSnapshot();
  });
});