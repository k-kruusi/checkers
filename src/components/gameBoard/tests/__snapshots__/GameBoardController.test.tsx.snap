// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoardController matches the snapshot 1`] = `
<DocumentFragment>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; margin: 20px 20px 20px 20px;"
  >
    <div
      style="display: grid; grid-template-columns: repeat(8, auto); gap: 0; border: 4px solid #daa520; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2); border-radius: 4px; padding: 4px;"
    >
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="false"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(231, 76, 60); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(34, 34, 34);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(70, 70, 70); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            draggable="true"
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455); background-color: rgb(34, 34, 34); border-radius: 50%; opacity: 1; display: flex; justify-content: center; align-items: center; font-weight: bold; color: rgb(255, 255, 240);"
          />
        </div>
      </div>
      <div>
        <div
          style="background-color: rgb(255, 255, 240); display: flex; justify-content: center; align-items: center; cursor: pointer; position: relative; padding: 4px;"
        >
          <div
            style="width: calc(100vw * 0.0455); height: calc(100vw * 0.0455);"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
